cli/tools/registry/pm/deps.rs

```diff
@@ -0,0 +1,969 @@
+      npm_fetch_resolver,
+      npm_resolver,
+      permissions_container,
+      main_module_graph_container,
+      lockfile,
+      pending_changes: Vec::new(),
+    }
+  }
+  pub fn from_workspace_dir(
+    workspace_dir: &Arc<WorkspaceDirectory>,
+    dep_filter: impl DepFilter,
+    args: DepManagerArgs,
+  ) -> Result<Self, AnyError> {
+    let mut deps = Vec::with_capacity(32);
+    if let Some(deno_json) = workspace_dir.maybe_deno_json() {
+      if deno_json.specifier.scheme() != "file" {
+        bail!("remote deno.json files are not supported");
+      }
+      let path = deno_json.specifier.to_file_path().unwrap();
+      if path.parent().unwrap() == workspace_dir.dir_path() {
+        add_deps_from_deno_json(deno_json, dep_filter, &mut deps);
+      }
+    }
+    if let Some(package_json) = workspace_dir.maybe_pkg_json() {
+      add_deps_from_package_json(package_json, dep_filter, &mut deps);
+    }
+
+    Ok(Self::with_deps_args(deps, args))
+  }
+  pub fn from_workspace(
+    workspace: &Arc<Workspace>,
+    dep_filter: impl DepFilter,
+    args: DepManagerArgs,
+  ) -> Result<Self, AnyError> {
+    let deps = deps_from_workspace(workspace, dep_filter)?;
+    Ok(Self::with_deps_args(deps, args))
+  }
+
+  async fn run_dependency_resolution(&self) -> Result<(), AnyError> {
+    if self
+      .dependencies_resolved
+      .load(std::sync::atomic::Ordering::Relaxed)
+    {
+      return Ok(());
+    }
+
+    let mut graph_permit = self
+      .main_module_graph_container
+      .acquire_update_permit()
+      .await;
+    let graph = graph_permit.graph_mut();
+    // populate the information from the lockfile
+    if let Some(lockfile) = &self.lockfile {
+      let lockfile = lockfile.lock();
+      graph.fill_from_lockfile(FillFromLockfileOptions {
+        redirects: lockfile
+          .content
+          .redirects
+          .iter()
+          .map(|(from, to)| (from.as_str(), to.as_str())),
+        package_specifiers: lockfile
+          .content
+          .packages
+          .specifiers
+          .iter()
+          .map(|(dep, id)| (dep, id.as_str())),
+      });
+    }
+
+    let npm_resolver = self.npm_resolver.as_managed().unwrap();
+    if self.deps.iter().all(|dep| match dep.kind {
+      DepKind::Npm => {
+        npm_resolver.resolve_pkg_id_from_pkg_req(&dep.req).is_ok()
+      }
+      DepKind::Jsr => graph.packages.mappings().contains_key(&dep.req),
+    }) {
+      self
+        .dependencies_resolved
+        .store(true, std::sync::atomic::Ordering::Relaxed);
+      graph_permit.commit();
+      return Ok(());
+    }
+
+    npm_resolver.ensure_top_level_package_json_install().await?;
+    let mut roots = Vec::new();
+    let mut info_futures = FuturesUnordered::new();
+    for dep in &self.deps {
+      if dep.location.is_deno_json() {
+        match dep.kind {
+          DepKind::Npm => roots.push(
+            ModuleSpecifier::parse(&format!("npm:/{}/", dep.req)).unwrap(),
+          ),
```

Comment:
Where do we acquire info for npm packages?

Also you might want to limit the concurrency here a bit - having unbounded amount of futures pinging registry servers might cause temporary rate limiting

Reply:
https://github.com/denoland/deno/blob/c71b3c2f6613ad11c47367a957052d889ef17a6b/cli/tools/registry/pm/deps.rs#L749

It's limited by a semaphore, currently 32 concurrent request. (though now that I'm looking, there should be separate limiting for jsr and npm, i'll fix that)